### PR TITLE
Cover Redis BlockQuery with tests

### PR DESCRIPTION
**Pull request by Warchant**
### Description of the Change

- [x] move `check_consistency`, `Identifier`, `id_to_name` to class FlatFile as static members, instead of free members. I did this at least because I needed `id_to_name` in `FlatFile` interface.
- [x] Fix bug in https://github.com/hyperledger/iroha/blob/develop/irohad/ametsuchi/impl/redis_block_query.cpp#L30-L35 . Test for bug + description is at https://github.com/hyperledger/iroha/compare/develop...feature/cover-redis-negative-cases?expand=1#diff-c855aee60870118507c2cd5177c150eaR241
- [x] Increase coverage of `BlockQuery`.

### Benefits

Bugfixes, increase coverage

### Possible Drawbacks 

None.

### CODE REVIEW

File: test/module/irohad/ametsuchi/block_query_test.cpp [hunk 1]
```diff
@@ -212,3 +215,154 @@ TEST_F(BlockQueryTest, GetTransactionsWithInvalidTxAndValidTx) {
...
+
+/**
+ * @given block store with 2 blocks totally containing 3 txs created by
+ * user1@test AND 1 tx created by user2@test
+ * @when get non-existent 1000 block
+ * @then nothing is returned
+ */
+TEST_F(BlockQueryTest, BlockQuery_GetBlocks_GetNonExistentBlock) {
+  auto wrapper = make_test_subscriber<CallExact>(
+      blocks->getBlocks(1000, 1), 0);
```
Comment (l4l) [inline test/module/irohad/ametsuchi/block_query_test.cpp]: Seem there's missing edge-case test (e.g getting 2nd blocks)

File: test/module/irohad/ametsuchi/block_query_test.cpp [hunk 2]
```diff
@@ -212,3 +215,154 @@ TEST_F(BlockQueryTest, GetTransactionsWithInvalidTxAndValidTx) {
   });
   ASSERT_TRUE(wrapper.validate());
 }
+
+/**
+ * @given block store with 2 blocks totally containing 3 txs created by
+ * user1@test AND 1 tx created by user2@test
```
Comment (l4l) [inline test/module/irohad/ametsuchi/block_query_test.cpp]: Imo as long as such details are inside the make_test_subscriber don't need to state it

File: test/module/irohad/ametsuchi/block_query_test.cpp [hunk 3]
```diff
@@ -212,3 +215,154 @@ TEST_F(BlockQueryTest, GetTransactionsWithInvalidTxAndValidTx) {
...
+ * @when get non-existent 1000 block
+ * @then nothing is returned
+ */
+TEST_F(BlockQueryTest, BlockQuery_GetBlocks_GetNonExistentBlock) {
```
Comment (l4l) [inline test/module/irohad/ametsuchi/block_query_test.cpp]: Maybe just `BlockQuery_GetNonExistentBlock`?

File: test/module/irohad/ametsuchi/block_query_test.cpp [hunk 4]
```diff
@@ -212,3 +215,154 @@ TEST_F(BlockQueryTest, GetTransactionsWithInvalidTxAndValidTx) {
...
+  ASSERT_TRUE(wrapper.validate());
+}
+
+/**
+ * @given block store with 2 blocks totally containing 3 txs created by
+ * user1@test AND 1 tx created by user2@test. Block #1 is filled with trash data (NOT JSON).
+ * @when read block #1
+ * @then get no blocks / error
+ */
+TEST_F(BlockQueryTest, BlockQuery_GetBlocks_GetBlockButItIsInvalidBlock) {
```
Comment (l4l) [inline test/module/irohad/ametsuchi/block_query_test.cpp]: Test name is getting huge

File: test/module/irohad/ametsuchi/block_query_test.cpp [hunk 5]
```diff
@@ -212,3 +215,154 @@ TEST_F(BlockQueryTest, GetTransactionsWithInvalidTxAndValidTx) {
...
+ * @given block store with 2 blocks totally containing 3 txs created by
+ * user1@test AND 1 tx created by user2@test
+ * @when get all blocks starting from 1
+ * @then returned all blocks (2)
+ */
+TEST_F(BlockQueryTest, BlockQuery_GetBlocksFrom_GetAllBlocksFrom1) {
+  auto wrapper = make_test_subscriber<CallExact>(
+      blocks->getBlocksFrom(1), this->blocks_total);
+  size_t counter = 1;
+  wrapper.subscribe([this, &counter](Block b) {
```
Comment (l4l) [inline test/module/irohad/ametsuchi/block_query_test.cpp]:
```
block_query_test.cpp:284:22: error: lambda capture 'this' is not used [-Werror,-Wunused-lambda-capture]
  wrapper.subscribe([this, &counter](Block b) {
                     ^
block_query_test.cpp:311:22: error: lambda capture 'this' is not used [-Werror,-Wunused-lambda-capture]
  wrapper.subscribe([this, block_n](Block block) {
                     ^
block_query_test.cpp:342:22: error: lambda capture 'this' is not used [-Werror,-Wunused-lambda-capture]
  wrapper.subscribe([this, block_n](Block block) {
                     ^
block_query_test.cpp:363:22: error: lambda capture 'this' is not used [-Werror,-Wunused-lambda-capture]
  wrapper.subscribe([this, &counter](Block b) {
```

File: test/module/irohad/ametsuchi/block_query_test.cpp (hunk 2 above)
Comment (Warchant) [reply test/module/irohad/ametsuchi/block_query_test.cpp]: These details are in fixture, not in wrapper. Every test specifies given, these are not exceptions

File: test/module/irohad/ametsuchi/block_query_test.cpp [hunk 6]
```diff
@@ -212,3 +215,154 @@ TEST_F(BlockQueryTest, GetTransactionsWithInvalidTxAndValidTx) {
...
+      blocks->getBlocks(1, 1), 1);
+  wrapper.subscribe();
+  ASSERT_TRUE(wrapper.validate());
+}
+
+/**
+ * @given block store with 2 blocks totally containing 3 txs created by
+ * user1@test AND 1 tx created by user2@test
+ * @when count=0
+ * @then returned empty block
```
Comment (muratovv) [inline test/module/irohad/ametsuchi/block_query_test.cpp]: "empty block" is very confusing. Better to write "return nothing" or similar.

File: test/module/irohad/ametsuchi/block_query_test.cpp [hunk 7]
```diff
@@ -212,3 +215,154 @@ TEST_F(BlockQueryTest, GetTransactionsWithInvalidTxAndValidTx) {
...
+      blocks->getBlocks(1, 0), 0);
+  wrapper.subscribe();
+  ASSERT_TRUE(wrapper.validate());
+}
+
+/**
+ * @given block store with 2 blocks totally containing 3 txs created by
+ * user1@test AND 1 tx created by user2@test
+ * @when get zero block
+ * @then returned empty block
```
Comment (muratovv) [inline test/module/irohad/ametsuchi/block_query_test.cpp]: Same issue as before.

File: test/module/irohad/ametsuchi/block_query_test.cpp [hunk 8]
```diff
@@ -212,3 +215,154 @@ TEST_F(BlockQueryTest, GetTransactionsWithInvalidTxAndValidTx) {
...
+  // write something that is NOT JSON to block #1
+  auto block_path = fs::path{this->block_store_path} / FlatFile::id_to_name(block_n);
+  fs::ofstream block_file(block_path);
+  std::string content = R"(this is definitely not json)";
+  block_file << content;
+  block_file.close();
+
+  auto wrapper = make_test_subscriber<CallExact>(
+      blocks->getBlocks(block_n, 1), 0);
+  wrapper.subscribe([this, block_n](Block block) {
```
Comment (muratovv) [inline test/module/irohad/ametsuchi/block_query_test.cpp]: What is the point of the lambda here? Maybe more clear just remove it?

File: test/module/irohad/ametsuchi/block_query_test.cpp (hunk 5 above)
Comment (Warchant) [reply test/module/irohad/ametsuchi/block_query_test.cpp]: What compiler?

File: test/module/irohad/ametsuchi/block_query_test.cpp (hunk 4 above)
Comment (Warchant) [reply test/module/irohad/ametsuchi/block_query_test.cpp]: But descriptive.

File: test/module/irohad/ametsuchi/block_query_test.cpp (hunk 1 above)
Comment (Warchant) [reply test/module/irohad/ametsuchi/block_query_test.cpp]: 1. there are blocks 1, 2
2. test name is "Get Non Existent Block". 1000 is more safe option.

File: test/module/irohad/ametsuchi/block_query_test.cpp (hunk 3 above)
Comment (Warchant) [reply test/module/irohad/ametsuchi/block_query_test.cpp]: Module_Method_TestCase

File: test/module/irohad/ametsuchi/block_query_test.cpp [hunk 9]
```diff
@@ -212,3 +215,154 @@ TEST_F(BlockQueryTest, GetTransactionsWithInvalidTxAndValidTx) {
...
+  std::string content = R"({
+  "testcase": [],
+  "description": "make sure this is valid json, but definitely not a block"
+})";
+  block_file << content;
+  block_file.close();
+
+  auto wrapper = make_test_subscriber<CallExact>(
+      blocks->getBlocks(block_n, 1), 0);
+  wrapper.subscribe([this, block_n](Block block) {
```
Comment (muratovv) [inline test/module/irohad/ametsuchi/block_query_test.cpp]: Same issue about lambda as before.

File: test/module/irohad/ametsuchi/block_query_test.cpp (hunk 3 above)
Comment (l4l) [reply test/module/irohad/ametsuchi/block_query_test.cpp]: Useless redundancy and it doesn't become more descriptive

File: test/module/irohad/ametsuchi/block_query_test.cpp (hunk 5 above)
Comment (l4l) [reply test/module/irohad/ametsuchi/block_query_test.cpp]: clang

File: test/module/irohad/ametsuchi/block_query_test.cpp (hunk 6 above)
Comment (l4l) [reply test/module/irohad/ametsuchi/block_query_test.cpp]: Agree, but empty block != nothing
maybe it's better "@then doesn't find such block"?

File: irohad/ametsuchi/impl/flat_file/flat_file.cpp [hunk 10]
```diff
@@ -178,3 +123,40 @@ FlatFile::FlatFile(Identifier current_id, const std::string &path)
   log_ = logger::log("FlatFile");
   current_id_.store(current_id);
 }
+
+const uint32_t FlatFile::DIGIT_CAPACITY = 16;
```
[the remark this answers is not in the document]
Comment (Warchant) [reply irohad/ametsuchi/impl/flat_file/flat_file.cpp]: It is not public. All functions but `id_to_name` are private.

[the remark this answers is not in the document]
Comment (Warchant) [reply irohad/ametsuchi/impl/flat_file/flat_file.cpp]: When you create a variable in cpp file -- by default it has external linkage. There will be problems if some other module defines the same constant.

[the remark this answers is not in the document]
Comment (Warchant) [reply irohad/ametsuchi/impl/flat_file/flat_file.cpp]: Also, `id_to_name` uses `Identifier`, which also should be accessible through FlatFile API.

File: test/module/irohad/ametsuchi/block_query_test.cpp (hunk 8 above)
Comment (Warchant) [reply test/module/irohad/ametsuchi/block_query_test.cpp]: Before the bugfix, lambda was executed. But now it is okay to remove it, I guess.

File: test/module/irohad/ametsuchi/block_query_test.cpp [hunk 11]
```diff
@@ -214,3 +217,146 @@ TEST_F(BlockQueryTest, GetTransactionsWithInvalidTxAndValidTx) {
...
+  ASSERT_TRUE(wrapper.validate());
+}
+
+/**
+ * @given block store with 2 blocks totally containing 3 txs created by
+ * user1@test AND 1 tx created by user2@test
+ * @when height=1, count=1
+ * @then returned exactly 1 block
+ *
+ * @note test for bug, when <total_blocks=2, height=1, count=1>, then
```
Comment (lebdron) [inline test/module/irohad/ametsuchi/block_query_test.cpp]: Redundant info about fixed bug in the code.

Comment (Warchant) [reply test/module/irohad/ametsuchi/block_query_test.cpp]: It is not redundant. Next developer may change the code and get the same bug, then he/she needs to understand what caused that bug.

Comment (lebdron) [reply test/module/irohad/ametsuchi/block_query_test.cpp]: Then there should be a comment about calculation in the following code, so that other developer will not change the code, but not in the test:
https://github.com/hyperledger/iroha/blob/97799acf1fc72c8e423dc8f36fe37f64f9f61c12/irohad/ametsuchi/impl/redis_block_query.cpp#L30-L34

File: irohad/ametsuchi/impl/flat_file/flat_file.cpp [hunk 12]
```diff
@@ -17,66 +17,26 @@
...
-  }  // namespace ametsuchi
-}  // namespace iroha
-
 using namespace iroha::ametsuchi;
+using Identifier = FlatFile::Identifier;
 
 // ----------| public API |----------
 
-nonstd::optional<std::unique_ptr<FlatFile>> FlatFile::create(const std::string &path) {
+std::string FlatFile::id_to_name(Identifier id) {
```
Comment (l4l) [inline irohad/ametsuchi/impl/flat_file/flat_file.cpp]: Why not `const Identifier&`?

File: test/module/irohad/ametsuchi/block_query_test.cpp [hunk 13]
```diff
@@ -209,8 +212,148 @@ TEST_F(BlockQueryTest, GetTransactionsWithInvalidTxAndValidTx) {
...
+}
+
+/**
+ * @given block store with 2 blocks totally containing 3 txs created by
+ * user1@test AND 1 tx created by user2@test. Block #1 is filled with trash data (NOT JSON).
+ * @when read block #1
+ * @then get no blocks
+ */
+TEST_F(BlockQueryTest, GetBlockButItIsInvalidBlock) {
+  namespace fs = boost::filesystem;
```
Comment (l4l) [inline test/module/irohad/ametsuchi/block_query_test.cpp]: Maybe extract to global space?

File: irohad/ametsuchi/impl/flat_file/flat_file.cpp (hunk 12 above)
Comment (Warchant) [reply irohad/ametsuchi/impl/flat_file/flat_file.cpp]: Identifier = uint32

File: test/module/irohad/ametsuchi/block_query_test.cpp (hunk 13 above)
Comment (Warchant) [reply test/module/irohad/ametsuchi/block_query_test.cpp]: Nop. Maybe in other tests I don't want to use fs as boost::filesystem?

File: irohad/ametsuchi/impl/flat_file/flat_file.cpp [hunk 14]
```diff
@@ -166,3 +126,39 @@ FlatFile::FlatFile(Identifier current_id,
...
+    log->error("check_consistency({}), not directory", dump_dir);
+    return nonstd::nullopt;
+  }
+
+  auto const files = [&dump_dir] {
+    std::vector<boost::filesystem::path> ps;
+    std::copy(boost::filesystem::directory_iterator{dump_dir},
+              boost::filesystem::directory_iterator{},
+              std::back_inserter(ps));
+    std::sort(ps.begin(),
```
Comment (l4l) [inline irohad/ametsuchi/impl/flat_file/flat_file.cpp]: Fix this statement as well plz: `std::sort(ps.begin(), ps.end(), std::less<boost::filesystem::path>());`

File: irohad/ametsuchi/impl/flat_file/flat_file.cpp (hunk 12 above)
Comment (l4l) [reply irohad/ametsuchi/impl/flat_file/flat_file.cpp]: oops, you're right

File: irohad/ametsuchi/impl/flat_file/flat_file.cpp (hunk 14 above)
Comment (lebdron) [reply irohad/ametsuchi/impl/flat_file/flat_file.cpp]: This seems out scope for this pull request, maybe create another task for it?
